### DISCUSSION TITLE
autocomplete doesn't work

### DISCUSSION THREAD

**Discussion by adsharma** [searxng/searxng#3511] [Q&A]
**Version of SearXNG, commit number if you are using on master branch and stipulate if you forked SearXNG**
2024.5.24+75e4b6512

**How did you install SearXNG?**
Docker container

**What happened?**
Added http://localhost:18888/ as a search engine in Chrome. Tried Firefox too. Same result.

Autocomplete works when I navigate to http://localhost:18888/ and start typing a query.
However, it does NOT work when I type the same query in the URL bar.

**How To Reproduce**

**Expected behavior**
Install searxng via docker and use the steps above.

**Screenshots & Logs**
n/a

**Additional context**
n/a

**Comment by adsharma** [comment]
Is this a bug or unsupported feature? I see that `opensearch.xml` has legit looking code to enable autocomplete in browsers. But then browser vendors stopped supporting `opensearch.xml`. So the only way to add a new search engine is to type the URL.

**Comment by cameronj86** [comment]
FML. After 4-5 months w/o autocorrect, I finally got it to work. I don't know what specifically did the trick though so I'm just documenting everything I tried over the last hour or two:

1. I opened up a dash shell (`docker exec -it searxng sh`) and checked the settings file to ensure that it is populated (`cat /etc/searxng/settings.yml`)
2. I then checked the environment (just type `env` when you are inside the container) and even tho AUTOCOMPLETE shows in the setting file, the variable is listed as empty.
3. I updated `docker-compose.yml` so that under the `environment:`  I manually state that `AUTOCOMPLETE=duckduckgo` (be sure to get the proper .yml indentation) **EDIT: Upon further investigation, this is _not_ the resolution: I spelled AUTOCOMPLETE wrong in my docker-compose file yet it worked lol**
4. I exited the interactive shell (Ctrl+D) and ran ~~docker compose restart~~ `docker compose down` and then `docker compose up -d` to make sure the variables where inherited.
5. I logged back into the interactive shell (`docker exec -it searxng sh`) and ran `env` and saw that `AUTOCOMPLETE` was populated 2x, and while one is still blank, another now shows duckduckgo.
6. **Note: This step will delete all the cookies that your browser associates w/ your server, so take a beat to just make sure all the passwords are saved beforehand.** I read somewhere that cached browser settings might play a role and conflict w/ updated container settings, so I opened my history and right clicked on a random searxng search and selected `Forget about this site` and deleted all cookies.
7. I _was_  using [libredirect](https://github.com/libredirect/browser_extension) as a hack-esque default search engine for my Firefox-based browser, but this time, I went to the searxng address, right clicked the address bar and selected 'add SearXNG' so that it was fully integrated w/ my browser.
8. I went to the searXNG preferences in my browser and made sure that the autocomplete option in the browser matches the variable I typed in my docker-compose file.

If I was a betting man, I'd say the most meaningful steps are 3, 6 and/or 7.

**Reply by cameronj86** [reply]
> FML. After 4-5 months w/o autocorrect, I finally got it to work. I don't know what specifically did the trick though so I'm just documenting everything I tried over the last hour or two:
> 
>     1. I opened up a dash shell (`docker exec -it searxng sh`) and checked the settings file to ensure that it is populated (`cat /etc/searxng/settings.yml`)
> 
>     2. I then checked the environment (just type `env` when you are inside the container) and even tho AUTOCOMPLETE shows in the setting file, the variable is listed as empty.
> 
>     3. I updated `docker-compose.yml` so that under the `environment:`  I manually state that `AUTOCOMPLETE=duckduckgo` (be sure to get the proper .yml indentation) **EDIT: Upon further investigation, this is _not_ the resolution: I spelled AUTOCOMPLETE wrong in my docker-compose file yet it worked lol**
> 
>     4. I exited the interactive shell (Ctrl+D) and ran ~docker compose restart~ `docker compose down` and then `docker compose up -d` to make sure the variables where inherited.
> 
>     5. I logged back into the interactive shell (`docker exec -it searxng sh`) and ran `env` and saw that `AUTOCOMPLETE` was populated 2x, and while one is still blank, another now shows duckduckgo.
> 
>     6. **Note: This step will delete all the cookies that your browser associates w/ your server, so take a beat to just make sure all the passwords are saved beforehand.** I read somewhere that cached browser settings might play a role and conflict w/ updated container settings, so I opened my history and right clicked on a random searxng search and selected `Forget about this site` and deleted all cookies.
> 
>     7. I _was_  using [libredirect](https://github.com/libredirect/browser_extension) as a hack-esque default search engine for my Firefox-based browser, but this time, I went to the searxng address, right clicked the address bar and selected 'add SearXNG' so that it was fully integrated w/ my browser.
> 
>     8. I went to the searXNG preferences in my browser and made sure that the autocomplete option in the browser matches the variable I typed in my docker-compose file.
> 
> 
> If I was a betting man, I'd say the most meaningful steps are 3, 6 and/or 7.

This broke recently after I changed ISPs. Not sure if it was luck of the draw or a holdover from prior tweaks, but the solution was much easier this time around:

1.  Delete any existing locally-hosted searxng search engine from the Firefox settings
2.  Set the Firefox default search to anything else
3. Go to your local edition and do a search
4. Right click the address bar and add your hosted version
5. When you switch FF back to searxng as your main engine, autocomplete should be working

**Comment by kchousos** [comment]
What worked for me was setting `autocomplete` to `duckduckgo` in the `/etc/searxng/settings.yml` file, like so:

```yaml
search:
  # Filter results. 0: None, 1: Moderate, 2: Strict
  safe_search: 0
  # Existing autocomplete backends: "dbpedia", "duckduckgo", "google", "yandex", "mwmbl",
  # "seznam", "startpage", "stract", "swisscows", "qwant", "wikipedia" - leave blank to turn it off
  # by default.
  autocomplete: "duckduckgo"
  ...
```

**Comment by epd5** [comment]
I'm getting this error any time I try to use autocomplete, tried all of the above suggestions, a fresh docker container with new settings & still not working. Running searxng/searxng docker container

> 2024-09-27 10:03:20,079 ERROR:searx.webapp: Exception on /autocompleter [POST]
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/webapp.py", line 849, in autocompleter
    for result in search_autocomplete(backend_name, sug_prefix, sxng_locale):
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/autocomplete.py", line 261, in search_autocomplete
    return backend(query, sxng_locale)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/autocomplete.py", line 174, in startpage
    data = resp.json()
           ^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/httpx/_models.py", line 756, in json
    return jsonlib.loads(self.text, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
[pid: 10|app: 0|req: 7/13] 172.19.0.25 () {62 vars in 1369 bytes} [Fri Sep 27 10:03:18 2024] POST /autocompleter => generated 265 bytes in 1680 msecs (HTTP/1.1 500) 8 headers in 291 bytes (1 switches on core 2)
2024-09-27 10:03:20,994 ERROR:searx.webapp: Exception on /autocompleter [POST]
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/webapp.py", line 849, in autocompleter
    for result in search_autocomplete(backend_name, sug_prefix, sxng_locale):
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/autocomplete.py", line 261, in search_autocomplete
    return backend(query, sxng_locale)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/autocomplete.py", line 174, in startpage
    data = resp.json()
           ^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/httpx/_models.py", line 756, in json
    return jsonlib.loads(self.text, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
[pid: 10|app: 0|req: 8/14] 172.19.0.25 () {62 vars in 1370 bytes} [Fri Sep 27 10:03:20 2024] POST /autocompleter => generated 265 bytes in 544 msecs (HTTP/1.1 500) 8 headers in 290 bytes (1 switches on core 1)

**Reply by allendema** [reply]
Looks* like your selected autocompleter was/is not working. Try another one.
